Add support for post-commit smart pointer workflow

briannafan opened the pull request:
#### fa1fd7865654668406db93da8758c67b96e9713d
<pre>
Add support for post-commit smart pointer workflow
<a href="https://bugs.webkit.org/show_bug.cgi?id=273896">https://bugs.webkit.org/show_bug.cgi?id=273896</a>
<a href="https://rdar.apple.com/127762312">rdar://127762312</a>

Reviewed by Jonathan Bedard.

Revamps the smart pointer static analysis scripts to find new regressions and
fixes as compared to expectation files per checker for each project.

* Tools/Scripts/compare-static-analysis-results:
(parser): Adds --check-expectations argument.
(find_diff): Removes grep commands and uses set operations instead.
(compare_project_results_to_expectations): New method to find unexpected regressions and fixes.
(compare_project_results_by_run): Renames method that finds new issues and fixes per run.
(main): Adds new logging.
(compare_project_results): Deleted.

* Tools/Scripts/generate-dirty-files:
(parse_results_file): Removes prefix from file paths.
(find_project_results): Changes saved data structure to lists of issues per file per checker.
(find_all_results): Creates issues_per_file.json which is consumed in compare-static-analysis-results.

* Tools/Scripts/generate-static-analysis-archive: Adds new templates for unexpected results page.
(parse_command_line):
(generate_results_page):
(create_results_file):
(main):

Canonical link: <a href="https://commits.webkit.org/278808@main">https://commits.webkit.org/278808@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4194f9ca758fe788c0d1bfdba3d88fe3522a439f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51543 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30853 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3895 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54810 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2236 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53846 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37182 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1916 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41962 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53642 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28492 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44453 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23088 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25793 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1714 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47757 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1800 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56402 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26662 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1671 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49365 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27902 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44523 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48556 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28796 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7528 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27637 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/compare-static-analysis-results:
```diff
@@ -24,22 +24,23 @@
+    parser.add_argument(
+        '--archived-dir',
+        dest='archived_dir',
+        help='directory of dirty lists from previous build'
```
While we're touching this, can we make it a formal sentance with capitalization and punctuation?

JonWBedard commented on Tools/Scripts/compare-static-analysis-results:
```diff
@@ -111,37 +159,39 @@ def compare_project_results(args, archive_path, new_path, project):
+    if unexpected_passes_total:
+        print(f'Total unexpected passing files: {len(unexpected_passes_total)}')
+    if unexpected_issues_total:
+        print(f'Total unexpected issues: {len(unexpected_issues_total)}')
```
I wonder if this would be more concise as a loop, something like:
```
for description, issues in {
'new issues': new_issues_total,
'unexpected failing files': unexpected_failures_total,
'unexpected passing files': unexpected_passes_total,
'unexpected issues': unexpected_issues_total,
}.items():
if lst:
print(f'Total {description}: {len(lst)}')
```

briannafan replied:
Oh I like this much more, thanks!

JonWBedard commented on Tools/Scripts/generate-static-analysis-archive:
```diff
@@ -63,8 +63,10 @@ def generate_results_page(project_dirs, id_string, output_root, static_analysis_
+    if result_type == 'new':
         heading = 'New issues in {}'.format(id_string)
+    elif result_type == 'unexpected':
+        heading = 'Unexpected issues from {}'.format(id_string)
```
Isn't this just `'{} issues from {}'.format(result_type, id_string)` in general?

briannafan replied:
Yes, thanks! I also applied the same logic change to `create_results_file`.

JonWBedard commented on Tools/Scripts/generate-static-analysis-archive:
```diff
@@ -112,7 +116,11 @@ def main(options):
-        create_results_file(options, output_root, static_analysis_new_dir, is_new_result=True)
+        create_results_file(options, output_root, static_analysis_new_dir, result_type='new')
+
+    if os.path.isdir(os.path.join(output_root, 'StaticAnalyzerUnexpectedRegressions')):
```
`StaticAnalyzerUnexpectedRegressions` should be a constant.

JonWBedard commented on Tools/Scripts/compare-static-analysis-results:
```diff
@@ -24,77 +24,125 @@
+            f.close()
+        if not os.path.exists(f'file2'):
+            f = open(f'file2', 'a')
+            f.close()
```
Why do we make `file1` and `file2` strings here?

briannafan replied:
Complete oversight 😅

JonWBedard commented on Tools/Scripts/compare-static-analysis-results:
```diff
@@ -24,77 +24,125 @@
+    path_to_reports = os.path.abspath(f'{args.build_output}/{category}/{project}/StaticAnalyzerReports')
+    subprocess.run(['mkdir', '-p', path_to_reports])
+    for issue_hash in issues:
+        report = f"report-{issue_hash[:6]}.html"
```
Shouldn't this be 12 characters? Why only 6?

briannafan replied:
`scan-build` uses the first 6 characters of an issue hash to name the report, like `report-7ab18e.html`

briannafan committed:
Add support for post-commit smart pointer workflow
https://bugs.webkit.org/show_bug.cgi?id=273896
rdar://127762312

Reviewed by Jonathan Bedard.

Revamps the smart pointer static analysis scripts to find new regressions and
fixes as compared to expectation files per checker for each project.

* Tools/Scripts/compare-static-analysis-results:
(parser): Adds --check-expectations argument.
(find_diff): Removes grep commands and uses set operations instead.
(compare_project_results_to_expectations): New method to find unexpected regressions and fixes.
(compare_project_results_by_run): Renames method that finds new issues and fixes per run.
(main): Adds new logging.
(compare_project_results): Deleted.

* Tools/Scripts/generate-dirty-files:
(parse_results_file): Removes prefix from file paths.
(find_project_results): Changes saved data structure to lists of issues per file per checker.
(find_all_results): Creates issues_per_file.json which is consumed in compare-static-analysis-results.

* Tools/Scripts/generate-static-analysis-archive: Adds new templates for unexpected results page.
(parse_command_line):
(generate_results_page):
(create_results_file):
(main):

Canonical link: https://commits.webkit.org/278808@main
